torch/csrc/jit/frontend/ir_emitter.cpp

```diff
@@ -1706,84 +1750,181 @@ struct to_ir {
   }
 
   CondValue emitIsInstance(const Expr& obj, const Expr& classinfo) {
-    // turn (float, (int, tuple)) into a flat list of types and type kind
-    // category checks: tuple_check = true, types = {float, int}
-    struct GatheredTypes {
-      GatheredTypes(ScriptTypeParser parser) : typeParser_(std::move(parser)) {}
-      void gather(const Expr& classinfo) {
-        if (classinfo.kind() == TK_TUPLE_LITERAL) {
-          for (Expr e : TupleLiteral(classinfo).inputs()) {
-            gather(e);
+    // For use with `unifyTypeList`
+    std::stringstream nowhere;
+
+    Value* lhs_val;
+
+    // Get the types on the right-hand side of the `isinstance`
+    // expression, e.g. `isinstance(x, (float, int))` -> `{float, int}.
+    // While doing this, we also want to decompose any `Union` types,
+    // namely `Union`, `Optional`, and `Number`
+    // NB: We can't use `auto` with this lambda because it contains a
+    // recursive call
+    std::function<void(const Expr&, std::vector<TypePtr>&, bool)> gather_types;
+
+    gather_types =
```

Comment:
The only shared logic here between lhs and rhs is the decomposition of Union types. I think this would be easier to follow if you factored that part out into its own function and then separately wrote out logic for lhs/rhs.